src/components/TextInput/TextInput.js

```diff
@@ -1,104 +1,190 @@
-        onClick(evt);
-      }
-    },
-    placeholder,
-    type,
+export default class TextInput extends React.Component {
+  state = {
+    type: this.props.type,
   };
 
-  const errorId = id + '-error-msg';
-  const textInputClasses = classNames('bx--text-input', className, {
-    'bx--text-input--light': light,
-  });
-  const labelClasses = classNames('bx--label', {
-    'bx--visually-hidden': hideLabel,
-  });
-
-  const label = labelText ? (
-    <label htmlFor={id} className={labelClasses}>
-      {labelText}
-    </label>
-  ) : null;
-
-  const error = invalid ? (
-    <div className="bx--form-requirement" id={errorId}>
-      {invalidText}
-    </div>
-  ) : null;
+  togglePasswordVisibility = () => {
+    this.setState({
+      type: this.state.type === 'password' ? 'text' : 'password',
+    });
+  };
 
-  const input = invalid ? (
-    <input
-      {...other}
-      {...textInputProps}
-      data-invalid
-      aria-invalid
-      aria-describedby={errorId}
-      className={textInputClasses}
-    />
-  ) : (
-    <input {...other} {...textInputProps} className={textInputClasses} />
-  );
+  togglePasswordVisibilityIconProps = () => {
+    const passwordIsVisible = this.state.type === 'text';
+    return {
+      alt: `${passwordIsVisible ? 'Hide' : 'Show'} password`,
+      name: `visibility-${passwordIsVisible ? 'off' : 'on'}`,
+      description: `${passwordIsVisible ? 'Hide' : 'Show'} password`,
+    };
+  };
 
-  const helper = helperText ? (
-    <div className="bx--form__helper-text">{helperText}</div>
-  ) : null;
+  render() {
+    const {
+      labelText,
+      className,
+      id,
+      placeholder,
+      type,
+      onChange,
+      onClick,
+      hideLabel,
+      invalid,
+      invalidText,
+      helperText,
+      light,
+      passwordVisibilityToggle,
+      ...other
+    } = this.props;
+    const errorId = id + '-error-msg';
+    const textInputClasses = classNames('bx--text-input', className, {
+      'bx--text-input--light': light,
+    });
+    const labelClasses = classNames('bx--label', {
+      'bx--visually-hidden': hideLabel,
+    });
+    const label = labelText ? (
+      <label htmlFor={id} className={labelClasses}>
+        {labelText}
+      </label>
+    ) : null;
+    const sharedTextInputProps = {
+      id,
+      onChange: evt => {
+        if (!other.disabled) {
+          onChange(evt);
+        }
+      },
+      onClick: evt => {
+        if (!other.disabled) {
+          onClick(evt);
+        }
+      },
+      placeholder,
+      type: passwordVisibilityToggle ? this.state.type : type,
+      className: textInputClasses,
+      ...other,
+    };
+    const invalidProps = {
+      'data-invalid': invalid,
+      'aria-invalid': invalid,
+      'aria-describedby': errorId,
+    };
+    const textInputProps = ({ invalid }) => ({
+      ...sharedTextInputProps,
+      ...(invalid ? invalidProps : {}),
+      ...{
+        'data-toggle-password-visibility':
```

Comment:
Seems that we don't need to create an object and spread it? Another question I have is what distinction you had in mind here vs. `sharedTextInputProps`, esp. given I see only one reference to that.